Refactor the browser class docs for xk6-browser (#722)
grafana/k6-docs: src/data/markdown/docs/30 xk6-browser/01 xk6-browser/01 Browser.md

excerpt: "xk6-browser: Browser Class"
---

A Browser is created via [browserType.launch([options])](/javascript-api/xk6-browser/browsertype/#browsertype-launch-options).

| Method                                                                                                                                                                         | Description                                                                                                                                           |
|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|

> | <nobr>[browser.close()](/javascript-api/xk6-browser/browser/close)</nobr>                                                                                                      | Closes the browser and all of its pages (if any were opened).                                                                                         |

Author: 🤦 I was going to check whether `nobr` works across all browsers, but then forgot. I'll change it to work with `div` and `CSS`.

I'll try to merge in main into #582 and then merge #582 into this branch so that I can use the PixelPoint tooltop component.

> | <nobr>[browser.close()](/javascript-api/xk6-browser/browser/close)</nobr>                                                                                                      | Closes the browser and all of its pages (if any were opened).                                                                                         |

Author: Resolved in: 10d88016f50f77c8f68354edf7b13aaaf7ee7e30

| <nobr>[browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)</nobr>                                                                                            | Indicates whether the [CDP](https://chromedevtools.github.io/devtools-protocol/) connection to the browser process is active or not.                  |
| <nobr><span title="Work in progress">🚧</span> [browser.newContext([options])](/javascript-api/xk6-browser/browser/newcontext/)</nobr>                                         | Creates and returns a new [BrowserContexts](/javascript-api/xk6-browser/browsercontext/).                                                             |
| <nobr><span title="Work in progress">🚧</span> [browser.newPage([options])](/javascript-api/xk6-browser/browser/newpage)</nobr>                                                | Creates a new [Page](/javascript-api/xk6-browser/page/) in a new [BrowserContexts](/javascript-api/xk6-browser/browsercontext/) and returns the page. |

> | <nobr><span title="Work in progress">[🚧](https://github.com/grafana/xk6-browser/issues/96)</span> [browser.on('disconnected')](/javascript-api/xk6-browser/browser/on)</nobr> | Detects events from the browser application.                                                                                                          |

Reviewer: This one is actually done, see https://github.com/grafana/xk6-browser/pull/268. The link you added is for `BrowserContext.on()` which is still pending, but that's for the `BrowserContext` page, not here.

Author: Resolved in: 9a2dc280ae6320fe8ea2b9265abeb57bffd90b84

excerpt: "xk6-browser: Browser Class"
---

> The `Browser` class is the entry point for all your tests, and it is what interacts with the actual internet browser via [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) (CDP). It manages:

Reviewer 2: ```suggestion
The `Browser` class is the entry point for all your tests and interacts with the actual Internet browser via [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) (CDP). It manages:
```

Reviewer: ?

```suggestion
The `Browser` class is the entry point for all your tests, and it is what interacts with the actual web browser process via [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) (CDP). It manages:
```

:smile:

Reviewer 2: 👍 Let's remove the "process"?

```suggestion
The `Browser` class is the entry point for all your tests, and it is what interacts with the actual web browser via [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) (CDP). It manages:
```

- [BrowserContext](/javascript-api/xk6-browser/browsercontext/)s which is where you can set a variety of attributes to control the behavior of pages;
- and [Page](/javascript-api/xk6-browser/page/)s which is where your rendered site is displayed.

> It can be created via `launch()` which is an exported function from `'k6/x/browser'`. Launching a new instance of `Browser` will also create a new browser process.

Reviewer 2: ```suggestion
A new Browser instance (hence a new browser process) can be created using the `launch()` function of the `'k6/x/browser'` module.
```

| [browser.close()](/javascript-api/xk6-browser/browser/close)                              | Closes the browser and all of its pages (if any were opened).                                                                                         |
| [browser.contexts()](/javascript-api/xk6-browser/browser/contexts)                        | Allows you to access all open [BrowserContexts](/javascript-api/xk6-browser/browsercontext/)s.                                                        |
| <BWIPT id="453"/> [browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)  | Indicates whether the [CDP](https://chromedevtools.github.io/devtools-protocol/) connection to the browser process is active or not.                  |
| <BWIPT/> [browser.newContext([options])](/javascript-api/xk6-browser/browser/newcontext/) | Creates and returns a new [BrowserContexts](/javascript-api/xk6-browser/browsercontext/).                                                             |

> | <BWIPT/> [browser.newPage([options])](/javascript-api/xk6-browser/browser/newpage)        | Creates a new [Page](/javascript-api/xk6-browser/page/) in a new [BrowserContexts](/javascript-api/xk6-browser/browsercontext/) and returns the page. |

Reviewer: I wonder if we should create issues to link these WIP icons to, and in general, always have a link in these cases.

Otherwise it's not clear what exactly isn't working, or what's missing for it to be fully working. It also makes it visible and plannable for us, as I'm currently not sure why these are WIP.

This is not a blocker for this PR, but WDYT?

Author: Yep, that is a little confusing. So the reason I've let them as WIP is because a few of the options are still WIP. Happy to create an issue though that encapsulates the issues with the options.

Author: Resolved in: 92c839a3765d609c2b43322595652554695f5381

> | <BWIPT/> [browser.newContext([options])](/javascript-api/xk6-browser/browser/newcontext/) | Creates and returns a new [BrowserContexts](/javascript-api/xk6-browser/browsercontext/).                                                             |
> | <BWIPT/> [browser.newPage([options])](/javascript-api/xk6-browser/browser/newpage)        | Creates a new [Page](/javascript-api/xk6-browser/page/) in a new [BrowserContexts](/javascript-api/xk6-browser/browsercontext/) and returns the page. |

Reviewer: ```suggestion
| <BWIPT/> [browser.newContext([options])](/javascript-api/xk6-browser/browser/newcontext/) | Creates and returns a new [BrowserContext](/javascript-api/xk6-browser/browsercontext/).                                                             |
| <BWIPT/> [browser.newPage([options])](/javascript-api/xk6-browser/browser/newpage)        | Creates a new [Page](/javascript-api/xk6-browser/page/) in a new [BrowserContext](/javascript-api/xk6-browser/browsercontext/) and returns the page. |
```

> | [browser.on('disconnected')](/javascript-api/xk6-browser/browser/on)                      | Detects events from the browser application.                                                                                                          |

Reviewer 2: ```suggestion
| [browser.on('disconnected')](/javascript-api/xk6-browser/browser/on) | Detects the disconnection event from the browser application. |
```

Author: ```suggestion
| [browser.on('disconnected')](/javascript-api/xk6-browser/browser/on) | Detects the disconnected event from the browser application. |
```

> 

Author: Resolved by: 409e6c8128edc317158fb612b419972aab453046

> A Browser is created via [browserType.launch([options])](/javascript-api/xk6-browser/browsertype/#browsertype-launch-options).

Author: WDYT?:

```
The `Browser` class is the entry point for all your tests, and it is what interacts with the actual internet browser via [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) (CDP). It manages:
- [BrowserContext](/javascript-api/xk6-browser/browsercontext/)s which is where you can set a variety of attributes to control the behavior of pages;
- and [Page](/javascript-api/xk6-browser/page/)s which is where your site is displayed.
```

| State | Method                                                                                         | Description                                                                                                     |
| - | ---------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |

> | ✅ | [browser.close()](/javascript-api/xk6-browser/browser/close)                                   | Closes the browser and all of its pages (if any were opened).                                                   |

Author: Yeah, good point. I was wondering how it would look if the correct UTF8 fonts weren't available. Is this something I can request now or at least start the conversation going?

> | ✅ | [browser.close()](/javascript-api/xk6-browser/browser/close)                                   | Closes the browser and all of its pages (if any were opened).                                                   |

Author: Created an issue to track this: https://github.com/grafana/k6-docs/issues/729

> | ✅ | [browser.contexts()](/javascript-api/xk6-browser/browser/contexts)                             | Allows you to access all open browser contexts.                                                                 |

Author: Resolved in: a2a0df4632445db8773fa9409a0514172e5c4475

> | ✅ | [browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)                         | Indicates whether the WebSocket connection to the browser application is active or not.                         |

Reviewer: Should we mention WebSockets at all here? The fact CDP is implemented over WS is an implementation detail, no?

```suggestion
| ✅ | [browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)                         | Indicates whether the [CDP](https://chromedevtools.github.io/devtools-protocol/) connection to the browser process is active or not.                         |
```

From a quick glance, Playwright mentions WS only in passing. I think we should do the same.

Author: Resolved in: 9304e039e9a18b78f25ecf3a0c2085201586faab

> | ✅ | [browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)                         | Indicates whether the WebSocket connection to the browser application is active or not.                         |

Author: Yeah, i was thinking the same thing when i decided to add the state column (will need to add an emoji to signify that it's implemented). The issue is with the display if there is no state column:

<img width="305" alt="Without state column" src="https://user-images.githubusercontent.com/1112428/178001708-f01b8f70-6870-43d4-bf03-8e2677e83fe6.png">

<img width="373" alt="With state column" src="https://user-images.githubusercontent.com/1112428/178001753-fbd069a0-65ea-4f44-8b40-98bbe94a73f8.png">

Happy to remove the state column if you prefer that still.

> | ✅ | [browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)                         | Indicates whether the WebSocket connection to the browser application is active or not.                         |

Reviewer: I prefer the look of the first screenshot, besides that line break for `newBrowserCDPSession()`. If we can avoid that by wrapping both the icon and link in a `div`, or using [`white-space: nowrap`](https://developer.mozilla.org/en-US/docs/Web/CSS/white-space), I think it would look OK.

I actually like that it offsets the method names, as it's immediately visible which methods might have problems, since they stick out from the rest of the fully implemented ones. Besides, once all methods are fully implemented, we'd presumably drop the "State" column altogether, so we'd end up with pages that have it, and pages that don't.

> | ✅ | [browser.isConnected](/javascript-api/xk6-browser/browser/isconnected)                         | Indicates whether the WebSocket connection to the browser application is active or not.                         |

Author: Resolved in: 19b13f672469cabf9508025f93a1e275046d7920

> | ❌ | [browser.newBrowserCDPSession()](/javascript-api/xk6-browser/browser/newbrowsercdpsession) | Allows you to access Browser's CDP session and talk directly to the browser application using the CDP protocol. |

Author: Yep, i agree with this. I will:

1. Remove rows of features which have a state of ❌ "Not implemented".
2. Remove the "Not implemented yet" page.
3. Replace the `MISSING PLAYWRIGHT APIS` column with a description column on the front page (issue #450).

> | ❌ | [browser.newBrowserCDPSession()](/javascript-api/xk6-browser/browser/newbrowsercdpsession) | Allows you to access Browser's CDP session and talk directly to the browser application using the CDP protocol. |

Reviewer: Sounds good, but I'm not clear what you mean in the 3rd point.

> | ❌ | [browser.newBrowserCDPSession()](/javascript-api/xk6-browser/browser/newbrowsercdpsession) | Allows you to access Browser's CDP session and talk directly to the browser application using the CDP protocol. |

Author: Here's a quick screenshot of what i mean by the 3rd point:

<img width="861" alt="Screenshot 2022-07-11 at 14 37 05" src="https://user-images.githubusercontent.com/1112428/178277425-c757d8d6-2f6e-450d-afdd-56765913365d.png">

> | ❌ | [browser.newBrowserCDPSession()](/javascript-api/xk6-browser/browser/newbrowsercdpsession) | Allows you to access Browser's CDP session and talk directly to the browser application using the CDP protocol. |

Author: Resolved in: 86b457ebb409f58828bacca1186bce545bfa1386

> | 🚧 | [browser.newContext([options])](/javascript-api/xk6-browser/browser/newcontext/)               | Creates and returns a new browser context.                                                                      |

Reviewer: ```suggestion
| 🚧 | [browser.newContext([options])](/javascript-api/xk6-browser/browser/newcontext/)               | Creates and returns a new [BrowserContext](/javascript-api/xk6-browser/browsercontext/).                                                                      |
```

Author: Resolved in: d68bc5c6987a93513be2151359ddc48033ac1512
